hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/SharedInstanceProfileCredentialsProvider.java

```diff
@@ -58,6 +71,84 @@ public static SharedInstanceProfileCredentialsProvider getInstance() {
     return INSTANCE;
   }
 
+  private AWSCredentials readCredentialsFromHDFS() {
+    try {
+      FileSystem fs = FileSystem.get(new Configuration());
+      BufferedReader br = new BufferedReader(new InputStreamReader(fs.open(s3crednetialPath)));
+      String accessKey = br.readLine();
+      String secretKey = br.readLine();
+      String token = br.readLine();
+      AWSCredentials credentials;
+      if (StringUtils.isEmpty(accessKey) || StringUtils.isEmpty(secretKey)) {
+        // if there are no accessKey nor secretKey return null
+        return null;
+      } else if (StringUtils.isNotEmpty(token)) {
+        credentials = new BasicSessionCredentials(accessKey, secretKey, token);
+      } else {
+        credentials = new BasicAWSCredentials(accessKey, secretKey);
+      }
+      return credentials;
+    } catch (Exception e) {
+      return null; // ignore the read errors
+      // throw new AmazonServiceException("Failed reading S3 credentials from HDFS " + e.getStackTrace());
+    }
+  }
+
+  private void writeCredentialsToHDFS(AWSCredentials credentials) {
+    try {
+      // Simulate atomic write by creating a new s3credential file with random string suffix and rename to s3crednetialPath
+      Path newS3crednetialPath = new Path(s3crednetialPath.toUri() + RandomStringUtils.randomAlphanumeric(8));
+      FileSystem fs = FileSystem.get(new Configuration());
+      BufferedWriter br = new BufferedWriter(new OutputStreamWriter(fs.create(newS3crednetialPath, true)));
+      String accessKey = credentials.getAWSAccessKeyId();
+      String secretKey = credentials.getAWSSecretKey();
+      String token = "";
+      if (credentials instanceof BasicSessionCredentials) {
```

Comment:
I would only allow session credentials to persist, so as to reduce risk of leakage of persistent secrets